libtransact/src/contract/archive/mod.rs

```diff
@@ -14,3 +14,553 @@
+//!     .expect("failed to load .scar file");
+//! ```
+
+mod error;
+
+use std::env::{split_paths, var_os};
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+use bzip2::read::BzDecoder;
+use glob::glob;
+use semver::{Version, VersionReq};
+use tar::Archive;
+
+pub use error::Error;
+
+/// Load default location(s) for .scar files.
+pub fn default_scar_paths() -> Vec<PathBuf> {
+    match var_os("SCAR_PATH") {
+        Some(paths) => split_paths(&paths).collect(),
+        None => vec![PathBuf::from("/usr/share/scar")],
+    }
+}
+
+/// The definition of a smart contract, including the contract's associated metadata and the bytes
+/// of the smart contract itself.
+pub struct SmartContractArchive {
+    /// The raw bytes of the WASM smart contract.
+    pub contract: Vec<u8>,
+    /// The metadata associated with this smart contract.
+    pub metadata: SmartContractMetadata,
+}
+
+impl SmartContractArchive {
+    /// Attempt to load a `SmartContractArchive` from a .scar file in the given `path`.
+    ///
+    /// The .scar file to load will be named `<name>_<version>.scar`, where `<name>` is the name
+    /// specified for this method and `<version>` is a valid semver string.
+    ///
+    /// The `version` argument for this method must be a valid semver requirement string. The
+    /// latest contract (based on semver requirement rules) that matches the given name and meets
+    /// the version requirements will be loaded.
+    ///
+    /// For more info on semver: https://semver.org/
+    ///
+    /// # Example
+    ///
+    /// If there are two files `/usr/share/scar/xo_1.2.3.scar` and `/usr/share/scar/xo_1.2.4`, you
+    /// can load v1.2.4 with:
+    ///
+    /// ```no_run
+    /// use transact::contract::archive::SmartContractArchive;
+    ///
+    /// let archive = SmartContractArchive::from_scar_file("xo", "1.2", "/usr/share/scar")
+    ///     .expect("failed to load .scar file");
+    /// assert_eq!(&archive.metadata.version, "1.2.4");
+    /// ```
+    ///
+    /// To load v1.2.3 explicitly, specify the exact version:
+    ///
+    /// ```no_run
+    /// use transact::contract::archive::SmartContractArchive;
+    ///
+    /// let archive = SmartContractArchive::from_scar_file("xo", "1.2.3", "/usr/share/scar")
+    ///     .expect("failed to load .scar file");
+    /// assert_eq!(&archive.metadata.version, "1.2.3");
+    /// ```
+    pub fn from_scar_file<P: AsRef<Path>>(
+        name: &str,
+        version: &str,
+        path: P,
+    ) -> Result<SmartContractArchive, Error> {
+        let file_path = find_scar(name, version, path)?;
+
+        let scar_file = File::open(file_path)
+            .map_err(|err| Error::new_with_source("failed to open file", err.into()))?;
+        let mut archive = Archive::new(BzDecoder::new(scar_file));
+        let archive_entries = archive
+            .entries()
+            .map_err(|err| Error::new_with_source("failed to read scar file", err.into()))?;
```

Comment:
Same as above this should include what file failed to be read.

Reply:
Updated